examples/dataflow-csvio/src/main/java/com/google/example/csvio/CSVRowUtils.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.google.example.csvio;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.sdk.values.Row;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+/** Helpers to process CSV data into its expected {@link Schema.Field} types. */
+class CSVRowUtils {
+
+  static Row csvLineToRow(CSVFormat csvFormat, String header, String line, Schema schema) {
+    try {
+      CSVParser headerParser = CSVParser.parse(header, csvFormat);
+      CSVParser lineParser = CSVParser.parse(line, csvFormat);
+
+      List<CSVRecord> headerRecordsRaw = headerParser.getRecords();
```

Comment:
The reason for downstream checks instead of bundling is that certain variables cannot be instantiated until previous variables are checked.  This is based on the fact that CSVRecord holds a list of "lines".

For example,

1) headerParser.getRecords() returns a List<CSVRecord> from which we need to check that there is only one header CSVRecord

2) Then, we need to evaluate the CSVRecord to check whether its field count matches the schema field count.

Reply:
Yes I understand that we need 2 checks for both header & line but my comment was about ordering of these checks as shown below:
```
 List<CSVRecord> headerRecordsRaw = headerParser.getRecords();

   if (headerRecordsRaw.size() != 1) {
        throw new IllegalArgumentException(
            String.format("header should be a single CSVRecord, got: %d", headerRecordsRaw.size()));
      }

      CSVRecord headerRecord = headerRecordsRaw.get(0);

 if (headerRecord.size() != schema.getFieldCount()) {
        throw new IllegalArgumentException(
            String.format(
                "mismatch of header size and schema size. header: %d vs schema: %d",
                headerRecord.size(), schema.getFieldCount()));
      }
      
      List<CSVRecord> lineRecordsRaw = lineParser.getRecords();
      if (lineRecordsRaw.size() != 1) {
        throw new IllegalArgumentException(
            String.format("line should be a single CSVRecord, got: %d", headerRecordsRaw.size()));
      }

      CSVRecord lineRecord = lineRecordsRaw.get(0);

      if (lineRecord.size() != schema.getFieldCount()) {
        throw new IllegalArgumentException(
            String.format(
                "mismatch of line size and schema size. line: %d vs schema: %d",
                lineRecord.size(), schema.getFieldCount()));
      }
```

Reply:
Ah! I understand now thank you.